src/internet/spotifywebapi/spotifywebapiservice.cpp

```diff
@@ -0,0 +1,189 @@
+      network_(new NetworkAccessManager{this}) {
+  app_->global_search()->AddProvider(
+      new SpotifyWebApiSearchProvider(app_, this));
+}
+
+SpotifyWebApiService::~SpotifyWebApiService() {}
+
+QStandardItem* SpotifyWebApiService::CreateRootItem() {
+  root_ = new QStandardItem(IconLoader::Load("spotify", IconLoader::Provider),
+                            kServiceName);
+  return root_;
+}
+
+void SpotifyWebApiService::LazyPopulate(QStandardItem* item) {}
+
+void SpotifyWebApiService::Search(int searchId, QString queryStr) {
+  if (std::chrono::system_clock::now().time_since_epoch() >= tokenExpiration_) {
+    QNetworkRequest request{
+        QUrl{"https://open.spotify.com/"
+             "get_access_token?reason=transport&productType=web_player"}};
+    request.setRawHeader("Accept-Encoding", "gzip");
+
+    QNetworkReply* reply = network_->get(request);
+    connect(reply, &QNetworkReply::finished, [=]() {
+      reply->deleteLater();
+
+      OnTokenReady(ParseJsonReplyWithGzip(reply), searchId, queryStr);
+    });
+  } else {
+    OnReadyToSearch(searchId, queryStr);
+  }
+}
+
+void SpotifyWebApiService::OnTokenReady(const QJsonDocument& json, int searchId,
+                                        QString queryStr) {
+  if (!json.isEmpty()) {
+    token_ = Get(json, "accessToken").toString();
+    tokenExpiration_ = std::chrono::milliseconds{static_cast<uint64_t>(
+        Get(json, "accessTokenExpirationTimestampMs").toDouble())};
+
+    qLog(Debug) << "Spotify API Token:" << token_;
+
+    OnReadyToSearch(searchId, queryStr);
+  }
+}
+
+void SpotifyWebApiService::OnReadyToSearch(int searchId, QString queryStr) {
+  qLog(Debug) << "Spotify API Searching: " << queryStr;
+
+  QNetworkRequest request{
+      QUrl{QString("https://api.spotify.com/v1/search?q=%1&type=track&limit=50")
+               .arg(queryStr.toHtmlEscaped())}};
+
+  request.setRawHeader("Accept-Encoding", "gzip");
+  request.setRawHeader("Authorization", ("Bearer " + token_).toUtf8());
+
+  QNetworkReply* reply = network_->get(request);
+  connect(reply, &QNetworkReply::finished, [=] {
+    reply->deleteLater();
+
+    BuildResultList(ParseJsonReplyWithGzip(reply), searchId);
+  });
+}
+
+void SpotifyWebApiService::BuildResultList(const QJsonDocument& json,
+                                           int searchId) {
+  QList<Song> result;
+
+  auto concat = [](const QJsonArray& array, const char* name) -> QString {
+    QStringList ret;
+    for (auto&& item : array) {
+      ret << Get(item, name).toString();
+    }
+    return ret.join(", ");
+  };
+
+  for (auto&& item : Get(json, "tracks", "items").toArray()) {
+    Song song;
+
+    song.set_albumartist(
+        concat(Get(item, "album", "artists").toArray(), "name"));
+    song.set_album(Get(item, "album", "name").toString());
+    song.set_artist(concat(Get(item, "artists").toArray(), "name"));
+    song.set_disc(Get(item, "disc_number").toInt());
+    song.set_length_nanosec(std::chrono::nanoseconds{
```

Comment:
There are some constants that might be easier to work with than C++'s verbose `chrono` in `timeconstants.h`.